### PR TITLE
NAS-108322 / 20.12 / Applications UI

### PR DESCRIPTION
**Pull request by dkmullen**
Screenshot from 2020-12-09 07-49-38

The status indicators in the next picture are not listening for changes, so are accurate only on refresh. 
Screenshot from 2020-12-09 07-49-26

Plex form pt 1
Screenshot from 2020-12-09 07-50-18

Plex form pt 2
Screenshot from 2020-12-09 07-50-29

Nextcloud form
Screenshot from 2020-12-09 07-50-47

Portals open from the button of a running release:
Screenshot from 2020-12-09 07-51-49
Screenshot from 2020-12-09 07-51-35

Shots from the wizard for creating a new chart release
Screenshot from 2020-12-07 15-14-35
Screenshot from 2020-12-07 15-15-41
Screenshot from 2020-12-07 15-17-21
Screenshot from 2020-12-07 15-20-21
Screenshot from 2020-12-07 15-20-45
Screenshot from 2020-12-07 15-21-14
Screenshot from 2020-12-07 15-21-31
Screenshot from 2020-12-07 15-21-45

Kubernetes settings
Screenshot from 2020-12-07 15-25-33

Actions menu
Screenshot from 2020-12-07 15-13-17

### CODE REVIEW

**Review by kmoore134 — CHANGES_REQUESTED** [review]
This is looking fantastic! My only real comment would be to perhaps remove the GPU configuration section for the time being. We'll document how to enable it on the CLI until we can get some cycles to do a proper UI selector, since a prop:value setup right now is confusing and doesn't seem to fit the nature of the feature.

File: src/app/pages/applications/catalog/catalog.component.ts [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+  }
+
+  constructor(private dialogService: DialogService,
+    private mdDialog: MatDialog, private translate: TranslateService,
+    private router: Router, private core: CoreService, private modalService: ModalService,
+    private appService: ApplicationsService, private sysGeneralService: SystemGeneralService) { }
+
+  ngOnInit(): void {
+    this.appService.getAllCatalogItems().subscribe(res => {
+      if (Object.keys(res[0].trains.test).length > 0) {
```
Comment (william-gr) [inline src/app/pages/applications/catalog/catalog.component.ts:64]: Should we switch to stable trains, @sonicaj  ?

File: src/app/pages/applications/chart-releases/chart-releases.component.ts [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+  selector: 'app-charts',
+  templateUrl: './chart-releases.component.html',
+  styleUrls: ['../applications.component.scss']
+})
+export class ChartReleasesComponent implements OnInit {
+  public chartItems = [];
+  private dialogRef: any;
+  public tempIcon = '/assets/images/ix-original.png';
+  public plexIcon = 'https://www.plex.tv/wp-content/uploads/2018/01/pmp-icon-1.png';
+  nextcloudIcon = 'https://cdn.rawgit.com/docker-library/docs/defa5ffc7123177acd60ddef6e16bddf694cc35f/nextcloud/logo.svg';
```
Comment (william-gr) [inline src/app/pages/applications/chart-releases/chart-releases.component.ts:27]: public var? ; )

File: src/app/pages/applications/catalog/catalog.component.ts (hunk 1 above)
Comment (sonicaj) [reply src/app/pages/applications/catalog/catalog.component.ts:64]: We should but can we please do that like at the very end when all the reviewers have approved ? I am building this branch and going to test it out before approving :) So would be easier if it is like this for now.

**Comment by william-gr** [comment]
@sonicaj is it possible to deploy the plex chart and then enable the GPU manually from a CLI command?

**Comment by sonicaj** [comment]
@william-gr it is possible to add/remove it later on

**Comment by william-gr** [comment]
@dkmullen I am seeing "New Chart Release" in screenshots, should that be "Deploy Docker Image" ?

File: src/app/helptext/apps/apps.ts [hunk 3]
```diff
@@ -0,0 +1,479 @@
...
+      },
+      action: T('Roll back'),
+
+      msg: T('Roll back '),
+      job: T('Rolling back...')
+    }
+  },
+
+  chartForm: {
+    title: T('New Chart Release'),
```
Comment (william-gr) [inline src/app/helptext/apps/apps.ts:108]: @sonicaj think of a better title for these two?

File: src/app/pages/applications/applications.service.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+  setReplicaCount(name: string, count: number) {
+    return this.ws.call('chart.release.scale', [name, { replica_count: count}]);
+  }
+
+  getPodConsoleChoices(name: string) {
+    return this.ws.call('chart.release.pod_console_choices', [name]);
+  }
+
+  getInterfaces() {
+    return this.ws.call('interface.query');
```
Comment (william-gr) [inline src/app/pages/applications/applications.service.ts:49]: Is every interface a possible choice, @sonicaj ?

Or should we provide a interface_choices?

File: src/app/helptext/apps/apps.ts (hunk 3 above)
Comment (sonicaj) [reply src/app/helptext/apps/apps.ts:108]: `Install Application` ?
And for chart release name `Application Name` ?

**Review by william-gr — APPROVED** [review]
Other than the last concerns, this is looking great. Nice job @dkmullen !

File: src/app/pages/applications/applications.service.ts (hunk 4 above)
Comment (sonicaj) [reply src/app/pages/applications/applications.service.ts:49]: I think we should, it should work as we use macvlan/bridge automatically but still better to have a separate method to ensure there is no issue. I'll update Dennis.

File: src/app/pages/applications/chart-releases/chart-releases.component.ts (hunk 2 above)
Comment (dkmullen) [reply src/app/pages/applications/chart-releases/chart-releases.component.ts]: Changed to private

File: src/app/helptext/apps/apps.ts (hunk 3 above)
Comment (dkmullen) [reply src/app/helptext/apps/apps.ts:108]: got it :)

**Review by kmoore134 — APPROVED** [review]

**Review by sonicaj — APPROVED** [review]
We need to hide static ip fields when dhcp is set but i think we can go with it as it is right now. Good job :)

**Pull request merged**